osquery/events/windows/usn_journal_reader.cpp

```diff
@@ -0,0 +1,879 @@
+
+  // export_bits will only push as many unsigned chars as are needed to
+  // represent
+  // a value, so we've no guarantee to sizes, but if bits are set above the 64th
+  // bit, it's an extended file id
+  if (buffer.size() <= sizeof(std::uint64_t)) {
+    file_id.Type = FileIdType;
+
+    buffer.resize(sizeof(std::uint64_t));
+    auto id_ptr = reinterpret_cast<const std::uint64_t*>(buffer.data());
+    file_id.FileId.QuadPart = *id_ptr;
+
+  } else {
+    /* NOTE(ww): osquery retains Windows 7 compatibility,
+       which means that we don't have access to the FILE_ID_128
+       member of FILE_ID_DESCRIPTOR's id union. We use GUID instead,
+       since it's the same width and the actual typing probably won't
+       matter for our purposes.
+     */
+    file_id.Type = ObjectIdType;
+    buffer.resize(sizeof(GUID));
+
+    memcpy(&file_id.ObjectId, buffer.data(), sizeof(GUID));
+  }
+}
+
+namespace USNParsers {
+bool GetUpdateSequenceNumber(USN& usn, const USN_RECORD* record) {
+  assert(record->MajorVersion != 4U);
+
+  switch (record->MajorVersion) {
+  case 2U:
+    usn = reinterpret_cast<const USN_RECORD_V2*>(record)->Usn;
+    return true;
+
+  case 3U:
+    usn = reinterpret_cast<const USN_RECORD_V3*>(record)->Usn;
+    return true;
+
+  default:
+    return false;
+  }
+}
+
+bool GetFileReferenceNumber(USNFileReferenceNumber& ref_number,
+                            const USN_RECORD* record) {
+  assert(record->MajorVersion != 4U);
+
+  switch (record->MajorVersion) {
+  case 2U: {
+    ref_number =
+        reinterpret_cast<const USN_RECORD_V2*>(record)->FileReferenceNumber;
+
+    return true;
+  }
+
+  case 3U: {
+    const auto& byte_array = reinterpret_cast<const USN_RECORD_V3*>(record)
+                                 ->FileReferenceNumber.Identifier;
+
+    boostmp::import_bits(ref_number,
+                         byte_array,
+                         byte_array + sizeof(FILE_ID_128::Identifier),
+                         0,
+                         false);
+
+    return true;
+  }
+
+  default:
+    return false;
+  }
+}
+
+bool GetParentFileReferenceNumber(USNFileReferenceNumber& ref_number,
+                                  const USN_RECORD* record) {
+  assert(record->MajorVersion != 4U);
+
+  switch (record->MajorVersion) {
+  case 2U: {
+    ref_number = reinterpret_cast<const USN_RECORD_V2*>(record)
+                     ->ParentFileReferenceNumber;
+    return true;
+  }
+
+  case 3U: {
+    const auto& byte_array = reinterpret_cast<const USN_RECORD_V3*>(record)
+                                 ->ParentFileReferenceNumber.Identifier;
+
+    boostmp::import_bits(ref_number,
+                         byte_array,
+                         byte_array + sizeof(FILE_ID_128::Identifier),
+                         0,
+                         false);
+
+    return true;
+  }
+
+  default:
+    return false;
+  }
+}
+
+bool GetTimeStamp(std::time_t& record_timestamp, const USN_RECORD* record) {
```

Comment:
Might be easier to read if we have `GetFieldsV2(const USN_RECORD_V2*rec)` and `GetFieldsV3(const USN_RECORD_V3* rec)` , etc. rather than having a function for each field and switching on MajorVersion?